### ISSUE TITLE
Strange Loss Curve

### ISSUE THREAD

**Issue by ZetangForward** [state-spaces/mamba#199]
Hi, I am training Mamba-1.4B on a downstream NLP task (Dialogue generation). I load Mamba model with 
```
from mamba_ssm.models.mixer_seq_simple import MambaLMHeadModel
model = MambaLMHeadModel.from_pretrained("state-spaces/mamba-1.4b")
``` 
and then I implement the optimizer and lr-scheduler as mentioned in the paper:

```
def configure_optimizers(self):
        betas = (self.exp_cfg.beta_1, self.exp_cfg.beta_2)
        optimizer = optim.Adam(
            self.model.parameters(), 
            lr=self.exp_cfg.peak_lr,
            weight_decay=self.exp_cfg.weight_decay, 
            betas=betas, 
            eps=self.exp_cfg.eps
        )
        
        def get_scheduler(optimizer, num_training_steps, warmup_steps, peak_lr, last_lr):

            def lr_lambda(current_step):
                if current_step < warmup_steps:
                    return current_step / warmup_steps
                progress = (current_step - warmup_steps) / (num_training_steps - warmup_steps)
                cosine_decay = 0.5 * (1 + math.cos(math.pi * progress))
                lr = (last_lr + (peak_lr - last_lr) * cosine_decay)
                return lr / peak_lr
            
            return torch.optim.lr_scheduler.LambdaLR(optimizer, lr_lambda)

        scheduler = get_scheduler(optimizer, self.exp_cfg.num_training_steps, self.exp_cfg.warmup_steps, self.exp_cfg.peak_lr, self.exp_cfg.last_lr)

        lr_scheduler = {
            'scheduler': scheduler,
            'name': 'custom_scheduler',
            'interval': 'step',  # Ensure learning rate updates per step
            'frequency': 1,  # Optional: If you want to make sure it updates every step
        }
        
        return {
            "optimizer": optimizer,
            "lr_scheduler": lr_scheduler,
        }
```
where I set 
```
num_training_steps: 20000
warmup_steps: 2000
peak_lr: 0.0002
last_lr: 0.00001
beta_1: 0.9
beta_2: 0.95
weight_decay: 0.1
```

However, the **loss curve is strange**:

![image](https://github.com/state-spaces/mamba/assets/123983104/64b2ed59-0a83-421d-aae2-85614afefc25)
![image](https://github.com/state-spaces/mamba/assets/123983104/e72fa7c5-992e-4989-9480-9ed622595abf)

I wonder, are the implementation of optimizer and lr-scheduler right?
And can you provide me any suggestion for fine-tuning Mamba-1.4B?

**Comment by tridao** [comment]
Does the same thing happen if you use a Transformer (.e.g Pythia 1.4B)?
You can also take a look at the [Trouble shotting](https://github.com/state-spaces/mamba/tree/main?tab=readme-ov-file#troubleshooting) section of the README.

**Comment by ZetangForward** [comment]
> Does the same thing happen if you use a Transformer (.e.g Pythia 1.4B)? You can also take a look at the [Troubleshooting](https://github.com/state-spaces/mamba/tree/main?tab=readme-ov-file#troubleshooting) section of the README.

Thx for replying, I am trying to figure out the issue within it.

**Comment by ZetangForward** [comment]
This problem is fixed, THX

**Comment by ZetangForward** [comment]
Hi, the training loss is more normal, but recently I have tried a few NLP generation tasks, and I have observed a phenomenon: **although the loss on the training set is gradually decreasing, the loss on the validation set will show an increasing trend.** 

Specifically,  I continue to fine-tune on the pre-trained mamba-2.8B ckpt， and one loss curve is shown below (continual generation on Story Generation task).

![image](https://github.com/state-spaces/mamba/assets/123983104/2f55ae09-cbb9-4c39-ac31-17a1dc25e668)
![image](https://github.com/state-spaces/mamba/assets/123983104/859c8257-6b61-4a22-9b0a-8b61441d23ab)

I'm curious if other researchers who are training mamba are experiencing this when training other (non-synthetic) downstream tasks.

My guess is that this could be influenced by the **Selection Mechanism** in Mamba,  as this would make it easier for mamba to learn the distribution on the training set (overfitting) and not as easy to generalize to the development set.
